apps/rig_inbound_gateway/lib/rig_inbound_gateway_web/session.ex

```diff
@@ -0,0 +1,226 @@
+  @session_timout_ms 3_600_000
+
+  # ---
+  def start(query_params, opts \\ []) do
+    GenServer.start(__MODULE__, %{query_params: query_params}, opts)
+  end
+
+  def recv_events(server, last_event_id) do
+    GenServer.call(server, {:recv_events, last_event_id || "first_event", 0}, 20_000)
+  end
+
+  # ---
+
+  # Server Callbacks
+
+  @impl true
+  def init(%{query_params: query_params}) do
+    # Init Subscriptions: Setup sulbscriptions for the token & also provided as query parameter
+    with {:ok, jwt_subs} <- Subscriptions.from_token(query_params["jwt"]),
+         {:ok, query_subs} <-
+           Map.get(query_params, "subscriptions") |> Subscriptions.from_json() do
+      subscriptions = Enum.uniq(jwt_subs ++ query_subs)
+
+      # We initially provide a welcome event...
+      send(self(), :welcome_event)
+      # ...register subscriptions...
+      send(self(), {:set_subscriptions, subscriptions})
+      # ..and schedule periodic refresh:
+      Process.send_after(self(), :refresh_subscriptions, @subscription_refresh_interval_ms)
+
+      # Init Session Timeout Validation:
+      Process.send_after(
+        self(),
+        :validate_session_timeout,
+        @session_timeout_validation_interval_ms
+      )
+
+      {:ok,
+       %{
+         subscriptions: [],
+         event_buffer: EventBuffer.new(@event_buffer_size),
+         session_valid_until: DateTime.add(DateTime.utc_now(), @session_timout_ms, :millisecond)
+       }}
+    else
+      {:error, %Subscriptions.Error{} = e} ->
+        {:stop, Exception.message(e)}
+    end
+  end
+
+  # ---
+  # Waits for events and replies them if available
+  @impl true
+  def handle_call({:recv_events, last_event_id, _}, from, state) do
+    state = %{
+      state
+      | session_valid_until: DateTime.add(DateTime.utc_now(), @session_timout_ms, :millisecond)
+    }
+
+    case EventBuffer.events_since(state.event_buffer, last_event_id) do
+      {:ok, [events: [], last_event_id: last_event_id]} ->
+        # There are no events, so we let the client wait @window_size_ms and check again:
+        Process.send_after(
+          self(),
+          {:recv_events, from, last_event_id, 1},
+          @window_size_ms
+        )
+
+        {:noreply, state}
+
+      {:ok, [events: events, last_event_id: last_event_id]} ->
+        {:reply,
+         %{
+           last_event_id: last_event_id,
+           events: Enum.map(events, fn x -> x.json end),
+           status: :ok
+         }, state}
+
+      {:no_such_event, [not_found_id: _not_found_id, last_event_id: last_event_id]} ->
+        # the event_id provided by the client was outdated - we send the newest event id
+        {:reply, %{last_event_id: last_event_id, events: [], status: :no_such_event}, state}
+    end
+  end
+
+  # Waits for events and replies them if available (same as above, but as handle_info)
+  @impl true
+  def handle_info({:recv_events, from, last_event_id, tries}, state) do
```

Comment:
shouldn't this function also update `session_valid_until`?

Reply:
yep